### PR TITLE
C++11 style code (noexcept, forwarding reference, move), and spurious wakeup

### PR DESCRIPTION
**Pull request by Fdhvdu**
you may need a semaphore to prevent spurious wakeup

### CODE REVIEW

File: daemon/boardd/queue.hpp [hunk 1]
```diff
@@ -2,64 +2,67 @@
...
 #ifndef _QUEUE_H_
 #define _QUEUE_H_
-#include <queue>
-#include <mutex>
-#include <condition_variable>
+#include<condition_variable>
+#include<mutex>
+#include<queue>
+#include<stdexcept> //runtime_error
+#include<utility> //forward, move
```
Comment (wens) [inline daemon/boardd/queue.hpp:9]: Please keep the space...

File: daemon/boardd/queue.hpp [hunk 2]
```diff
@@ -2,64 +2,67 @@
...
+		cv_poped_.notify_one();
+		return true;
+	}
+	void Close()
+	{
+		std::unique_lock<std::mutex> lk{mu_};
+		eof_=true;
+		cv_pushed_.notify_all();
+		cv_poped_.notify_all();
+	}
```
Comment (wens) [inline daemon/boardd/queue.hpp:65]: Please don't reindent the whole file. It makes changes less obvious, and doesn't play nice with blame either.
Use whatever code style the original file had.

Comment (Fdhvdu) [inline daemon/boardd/queue.hpp:65]: How do I fix it?

File: daemon/boardd/queue.hpp [hunk 3]
```diff
@@ -2,64 +2,67 @@
...
+		std::unique_lock<std::mutex> lk{mu_};
+		while(true)
+		{
+			if(eof_)
+				throw std::runtime_error{"Queue closed"};
+			if(max_&&qu_.size()>=max_)
+			{
+				cv_poped_.wait(lk); //do you aware spurious wakeup?
+				                    //if not, you may need a semaphore
+				continue;
```
Comment (robertabcd) [inline daemon/boardd/queue.hpp:38]: This continue would do the check, wouldn't it? (L25 on the left side)

File: daemon/boardd/queue.hpp [hunk 4]
```diff
@@ -2,64 +2,67 @@
...
-  std::condition_variable cv_poped_;
-  bool eof_;
-  size_t max_;
+	Queue() noexcept
+		:Queue{}{}
+	Queue(std::size_t max) noexcept
+		:eof_{false},max_{max}{}
+	//probably need enable_if and is_convertible
+	template<class T2>
+	void Push(T2 &&elem)
```
Comment (robertabcd) [inline daemon/boardd/queue.hpp:27]: Why another template here?

File: daemon/boardd/queue.hpp [hunk 5]
```diff
@@ -2,64 +2,67 @@
...
+	}
+	bool Pop(T &elem)
+	{
+		std::unique_lock<std::mutex> lk{mu_};
+		while(qu_.empty())
+		{
+			if(eof_)
+				return false;
+			cv_pushed_.wait(lk);  //again, how about spurious wakeup?
+		}
```
Comment (robertabcd) [inline daemon/boardd/queue.hpp:53]: This while loop would ensure this, wouldn't it?
Please do not just raise problems in comment unless it's something that you can't fix.

File: daemon/boardd/queue.hpp (hunk 2 above)
Comment (robertabcd) [inline daemon/boardd/queue.hpp:65]: Yeah, please preserve the original coding style.

File: daemon/boardd/queue.hpp (hunk 3 above)
Comment (Fdhvdu) [inline daemon/boardd/queue.hpp:38]: Yes. You are right.
Even if there is a spurious wakeup, it is safe.

File: daemon/boardd/queue.hpp (hunk 5 above)
Comment (Fdhvdu) [inline daemon/boardd/queue.hpp:53]: Yes. You are right.
The while loop ensure it.

File: daemon/boardd/queue.hpp (hunk 4 above)
Comment (Fdhvdu) [inline daemon/boardd/queue.hpp:27]: That is for distinguish l-value and r-value.

**Pull request closed without merging**